src/sentry/runner/commands/devservices.py

```diff
@@ -553,3 +551,97 @@ def rm(project: str, services: list[str]) -> None:
             else:
                 click.secho("> Removing '%s' network" % network.name, err=True, fg="red")
                 network.remove()
+
+
+def check_health(service_name: str, containers: dict[str, Any]) -> None:
+    healthcheck = getattr(service_healthchecks, service_name, None)
+    if healthcheck is None:
+        return
+
+    click.secho(f"> Checking container health '{service_name}'", fg="yellow")
+
+    def hc() -> None:
+        if healthcheck is None:
+            return
+        healthcheck.check(containers)
+
+    try:
+        run_with_retries(hc)
+    except subprocess.CalledProcessError:
+        click.secho(f"> '{service_name}' is not healthy", fg="red")
+        raise
+
+
+def run_with_retries(cmd: Callable[[], object], retries: int = 3, timeout: int = 5) -> None:
+    for retry in range(1, retries + 1):
+        try:
+            cmd()
+        except (subprocess.CalledProcessError):
+            if retry == retries:
+                raise
+            else:
+                click.secho(
+                    f"  > Health check failed, retrying in {timeout}s (attempt {retry+1} of {retries})...",
+                    fg="yellow",
+                )
+                time.sleep(timeout)
+        else:
+            return
+
+
+def check_kafka(containers: dict[str, Any]) -> None:
+    kafka_options = containers["kafka"]
+    zk_options = containers["zookeeper"]
+
+    zk_port = zk_options.get("environment", {}).get("ZOOKEEPER_CLIENT_PORT", 2181)
+
+    subprocess.run(
+        (
+            "docker",
+            "exec",
+            kafka_options["name"],
+            "kafka-topics",
+            "--zookeeper",
+            f"{zk_options['name']}:{zk_port}",
+            "--list",
+        ),
+        check=True,
+        capture_output=True,
+        text=True,
+    )
+
+
+def check_postgres(containers: dict[str, Any]) -> None:
+    pg_options = containers["postgres"]
+    subprocess.run(
+        (
+            "docker",
+            "exec",
+            pg_options["name"],
+            "pg_isready",
+            "-U",
+            "postgres",
+        ),
+        check=True,
+        capture_output=True,
+        text=True,
+    )
+
+
+class ServiceHealthcheck:
+    check: Callable[[dict[str, Any]], None]
+
+    def __init__(self, check: Callable[[dict[str, Any]], None]) -> None:
+        self.check = check
+
+
+class ServiceHealthChecks:
+    postgres: ServiceHealthcheck
+    kafka: ServiceHealthcheck
+
+    def __init__(self) -> None:
+        self.postgres = ServiceHealthcheck(check_postgres)
+        self.kafka = ServiceHealthcheck(check_kafka)
```

Comment:
this one was probably fine to leave as a dictionary -- since you actually _do_ want to map from the string to the useful bits